internal/runners/packages/packages.go

```diff
@@ -3,49 +3,79 @@ package packages
+	"github.com/go-openapi/strfmt"
 )
 
 const latestVersion = "latest"
 
 func executePackageOperation(out output.Outputer, prompt prompt.Prompter, language, name, version string, operation model.Operation) error {
-	fail := auth.RequireAuthentication(locale.T("auth_required_activate"), out, prompt)
-	if fail != nil {
-		return fail.WithDescription("err_activate_auth_required")
+	isHeadless := false
+	if !authentication.Get().Authenticated() {
+		anonymousOk, fail := prompt.Confirm(locale.T("prompt_headless_anonymous"), true)
+		if fail != nil {
+			return locale.WrapInputError(fail.ToError(), "Failed to determine if it is okay to proceed without authorization.")
+		}
+		isHeadless = anonymousOk
+	}
+
+	// Note: User also lands here if answering No to the question about anonymous commit.
+	if !isHeadless {
+		fail := auth.RequireAuthentication(locale.T("auth_required_activate"), out, prompt)
+		if fail != nil {
+			return fail.WithDescription("err_activate_auth_required")
+		}
 	}
 
 	if strings.ToLower(version) == latestVersion {
 		version = ""
 	}
 
 	// Verify that the provided package actually exists (the vcs API doesn't care)
-	var ingredient *model.IngredientAndVersion
+	ingredientNamespace := ""
 	var err error
-	if version == "" {
-		ingredient, err = model.IngredientWithLatestVersion(language, name)
-	} else {
-		ingredient, err = model.IngredientByNameAndVersion(language, name, version)
-	}
-	if err != nil {
-		return locale.WrapError(err, "package_ingredient_err", "Failed to resolve an ingredient named {{.V0}}.", name)
+	if operation != model.OperationRemoved {
+		var ingredient *model.IngredientAndVersion
+		if version == "" {
+			ingredient, err = model.IngredientWithLatestVersion(language, name)
+		} else {
+			ingredient, err = model.IngredientByNameAndVersion(language, name, version)
+		}
+		if err != nil {
+			return locale.WrapError(err, "package_ingredient_err", "Failed to resolve an ingredient named {{.V0}}.", name)
+		}
+		ingredientNamespace = ingredient.Namespace
 	}
 
-	// Commit the package
 	pj := project.Get()
-	commitID, fail := model.CommitPackage(pj.Owner(), pj.Name(), operation, name, ingredient.Namespace, version)
-	if fail != nil {
-		return fail.WithDescription("err_package_" + string(operation)).ToError()
+	var commitID strfmt.UUID
+	if isHeadless {
+		parentCommitID, err := pj.CommitUUID()
+		if err != nil {
+			return locale.WrapError(err, "package_headless_invalid_commit_id", "Failed to determine current commit.")
+		}
+
+		var fail *failures.Failure
+		commitID, fail = model.CommitPackage(parentCommitID, operation, name, ingredientNamespace, version)
+		if fail != nil {
+			return locale.WrapError(fail.ToError(), "err_package_"+string(operation))
+		}
+	} else {
+		var fail *failures.Failure
+		commitID, fail = model.CommitPackageInBranch(pj.Owner(), pj.Name(), operation, ingredientNamespace, name, version)
+		if fail != nil {
+			return fail.WithDescription("err_package_" + string(operation)).ToError()
+		}
```

Comment:
Let's clean this up. We should have one model function that creates the commit, and another model function that updates the branch. Both should be called from here, but updating the branch should only be called if `!isHeadless`.